mmdet/models/dense_heads/ascend_ssd_head.py

```diff
@@ -0,0 +1,242 @@
+            reg_decoded_bbox,
+            train_cfg,
+            test_cfg,
+            init_cfg)
+
+    def get_static_anchors(self, featmap_sizes, img_metas, device="cuda"):
+        if not hasattr(self, 'static_anchors') or not hasattr(self, 'static_valid_flags'):
+            static_anchors, static_valid_flags = self.get_anchors(featmap_sizes, img_metas, device)
+            self.static_anchors = static_anchors
+            self.static_valid_flags = static_valid_flags
+        return self.static_anchors, self.static_valid_flags
+
+    def get_targets(self,
+                    anchor_list,
+                    valid_flag_list,
+                    gt_bboxes_list,
+                    img_metas,
+                    gt_bboxes_ignore_list=None,
+                    gt_labels_list=None,
+                    label_channels=1,
+                    unmap_outputs=True,
+                    return_sampling_results=False,
+                    return_level=True):
+        return AscendAnchorHead.get_targets(
+            self,
+            anchor_list,
+            valid_flag_list,
+            gt_bboxes_list,
+            img_metas,
+            gt_bboxes_ignore_list,
+            gt_labels_list,
+            label_channels,
+            unmap_outputs,
+            return_sampling_results,
+            return_level,
+        )
+
+    def concat_loss(self,
+                    concat_cls_score, concat_bbox_pred,
+                    concat_anchor, concat_labels,
+                    concat_label_weights,
+                    concat_bbox_targets, concat_bbox_weights,
+                    concat_pos_mask, concat_neg_mask,
+                    num_total_samples):
+        num_images, num_anchors, _ = concat_anchor.size()
+
+        concat_loss_cls_all = F.cross_entropy(
+            concat_cls_score.view((-1, self.cls_out_channels)), concat_labels.view(-1),
+            reduction='none').view(concat_label_weights.size()) * concat_label_weights
+        # # FG cat_id: [0, num_classes -1], BG cat_id: num_classes
+        concat_num_pos_samples = torch.sum(concat_pos_mask, dim=1)
+        concat_num_neg_samples = self.train_cfg.neg_pos_ratio * concat_num_pos_samples
+
+        concat_num_neg_samples_max = torch.sum(concat_neg_mask, dim=1)
+        concat_num_neg_samples = torch.min(concat_num_neg_samples, concat_num_neg_samples_max)
+
+        concat_topk_loss_cls_neg, _ = torch.topk(concat_loss_cls_all * concat_neg_mask, k=num_anchors, dim=1)
+        concat_loss_cls_pos = torch.sum(concat_loss_cls_all * concat_pos_mask, dim=1)
+
+        anchor_index = torch.arange(end=num_anchors, dtype=torch.float, device=concat_anchor.device).view((1, -1))
+        topk_loss_neg_mask = (anchor_index < concat_num_neg_samples.view(-1, 1)).float()
+
+        concat_loss_cls_neg = torch.sum(concat_topk_loss_cls_neg * topk_loss_neg_mask, dim=1)
+        loss_cls = (concat_loss_cls_pos + concat_loss_cls_neg) / num_total_samples
+
+        if self.reg_decoded_bbox:
```

Comment:
if self.reg_decoded_bbox should be False, how about add a assert in init?

Reply:
Solved